lib/Builtins.h

```diff
@@ -18,115 +18,160 @@
-inline bool IsUnsampledImageRead(llvm::Function *f) {
-  return IsUnsampledImageRead(f->getName());
-}
-
-// Returns true if the function is an OpenCL image read of float type.
-bool IsFloatUnsampledImageRead(llvm::StringRef name);
-inline bool IsFloatUnsampledImageRead(llvm::Function *f) {
-  return IsFloatUnsampledImageRead(f->getName());
-}
-
-// Returns true if the function is an OpenCL image read of uint type.
-bool IsUintUnsampledImageRead(llvm::StringRef name);
-inline bool IsUintUnsampledImageRead(llvm::Function *f) {
-  return IsUintUnsampledImageRead(f->getName());
-}
-
-// Returns true if the function is an OpenCL image read of int type.
-bool IsIntUnsampledImageRead(llvm::StringRef name);
-inline bool IsIntUnsampledImageRead(llvm::Function *f) {
-  return IsIntUnsampledImageRead(f->getName());
-}
-
-// Returns true if the function is an OpenCL image write.
-bool IsImageWrite(llvm::StringRef name);
-inline bool IsImageWrite(llvm::Function *f) {
-  return IsImageWrite(f->getName());
-}
-
-// Returns true if the function is an OpenCL image write of float type.
-bool IsFloatImageWrite(llvm::StringRef name);
-inline bool IsFloatImageWrite(llvm::Function *f) {
-  return IsFloatImageWrite(f->getName());
-}
-
-// Returns true if the function is an OpenCL image write of uint type.
-bool IsUintImageWrite(llvm::StringRef name);
-inline bool IsUintImageWrite(llvm::Function *f) {
-  return IsUintImageWrite(f->getName());
-}
-
-// Returns true if the function is an OpenCL image write of int type.
-bool IsIntImageWrite(llvm::StringRef name);
-inline bool IsIntImageWrite(llvm::Function *f) {
-  return IsIntImageWrite(f->getName());
-}
-
-// Returns true if the function is an OpenCL image height query.
-bool IsGetImageHeight(llvm::StringRef name);
-inline bool IsGetImageHeight(llvm::Function *f) {
-  return IsGetImageHeight(f->getName());
-}
-
-// Returns true if the function is an OpenCL image width query.
-bool IsGetImageWidth(llvm::StringRef name);
-inline bool IsGetImageWidth(llvm::Function *f) {
-  return IsGetImageWidth(f->getName());
-}
-
-// Returns true if the function is an OpenCL image depth query.
-bool IsGetImageDepth(llvm::StringRef name);
-inline bool IsGetImageDepth(llvm::Function *f) {
-  return IsGetImageDepth(f->getName());
-}
-
-// Returns true if the function is an OpenCL image dim query.
-bool IsGetImageDim(llvm::StringRef name);
-inline bool IsGetImageDim(llvm::Function *f) {
-  return IsGetImageDim(f->getName());
-}
-
-// Returns true if the function is an OpenCL image query.
-bool IsImageQuery(llvm::StringRef name);
-inline bool IsImageQuery(llvm::Function *f) {
-  return IsImageQuery(f->getName());
-}
+  namespace Builtins {
+
+    struct ParamTypeInfo {
+      bool                     sign               = false;                    // is signed
+      llvm::Type::TypeID       type               = llvm::Type::VoidTyID;     // elem type
+      int                      blen               = 0;                        // byte len
```

Comment:
Don't abbreviate as much:
 I'd like to see these renamed to  byte_len and vector_size (and then comment the latter that it's the number of elements)